File: frontend/src/api/dictionaries.ts
import { fetchWithAuth, parseJSONResponse } from './promo';

const API_BASE = import.meta.env.VITE_API_BASE || 'http://localhost:8080';

export interface SKUReference {
  id: number; sku: string; brand: string; brand_as: string; created_at: string;
}
export interface NetworkReference {
  id: number; network_name: string; kam: string; network_type: string; top20_segment: string; key_region: string;
}
export interface KAMNetworkReference {
  id: number; kam: string; network_name: string; valid_from: string; created_at: string;
}
export interface MechanicReference {
  id: number; mechanics: string; channel: string; short_code: string; created_at: string;
}
export interface DictionaryData {
  skus: SKUReference[];
  networks: NetworkReference[];
  kam_networks: KAMNetworkReference[];
  mechanics: MechanicReference[];
}

export type DictionaryKind = 'skus' | 'networks' | 'kam-networks' | 'mechanics';
export type DictionaryRow = SKUReference | NetworkReference | KAMNetworkReference | MechanicReference;

async function save<T extends DictionaryRow>(kind: DictionaryKind, data: Partial<T>): Promise<T> {
  const id = data.id;
  const response = await fetchWithAuth(
    `${API_BASE}/api/admin/dictionaries/${kind}${id ? `/${id}` : ''}`,
    { method: id ? 'PATCH' : 'POST', body: JSON.stringify(data) },
  );
  return parseJSONResponse<T>(response, 'Не удалось сохранить запись');
}

export const dictionariesAPI = {
  getAll: (): Promise<DictionaryData> =>
    fetchWithAuth(`${API_BASE}/api/admin/dictionaries`)
      .then(response => parseJSONResponse<DictionaryData>(response, 'Не удалось загрузить справочники')),
  save,
};

File: backend/models/dictionaries.go
package models

// SKUReference — соответствие коммерческого SKU брендам в источнике и АС.
type SKUReference struct {
	ID        int    `json:"id"`
	SKU       string `json:"sku"`
	Brand     string `json:"brand"`
	BrandAS   string `json:"brand_as"`
	CreatedAt string `json:"created_at"`
}

// NetworkReference — атрибуты сети, используемые промо и аналитикой продаж.
type NetworkReference struct {
	ID           int    `json:"id"`
	NetworkName  string `json:"network_name"`
	KAM          string `json:"kam"`
	NetworkType  string `json:"network_type"`
	Top20Segment string `json:"top20_segment"`
	KeyRegion    string `json:"key_region"`
}

// KAMNetworkReference — период действия закрепления КАМ за сетью.
type KAMNetworkReference struct {
	ID          int    `json:"id"`
	KAM         string `json:"kam"`
	NetworkName string `json:"network_name"`
	ValidFrom   string `json:"valid_from"`
	CreatedAt   string `json:"created_at"`
}

// MechanicReference — механика промо и её аналитический канал.
type MechanicReference struct {
	ID        int    `json:"id"`
	Mechanics string `json:"mechanics"`
	Channel   string `json:"channel"`
	ShortCode string `json:"short_code"`
	CreatedAt string `json:"created_at"`
}

type DictionaryData struct {
	SKUs        []SKUReference        `json:"skus"`
	Networks    []NetworkReference    `json:"networks"`
	KAMNetworks []KAMNetworkReference `json:"kam_networks"`
	Mechanics   []MechanicReference   `json:"mechanics"`
}
